lower.ts

```diff
@@ -291,6 +344,101 @@ function flattenExprToExpr(e : AST.Expr<Annotation>, env : GlobalEnv) : [Array<I
         ],
         { a: e.a, tag: "value", value: { a: e.a, tag: "id", name: newName } }
       ];
+    case "construct-list":
+      const newListName = generateName("newList");
+      const listAlloc: IR.Expr<Annotation> = {
+        tag: "alloc",
+        amount: { tag: "wasmint", value: e.items.length + 1 },
+      };
+      var inits: Array<IR.VarInit<Annotation>> = [];
+      var stmts: Array<IR.Stmt<Annotation>> = [];
+      var storeLength: IR.Stmt<Annotation> = {
+        tag: "store",
+        start: { tag: "id", name: newListName },
+        offset: { tag: "wasmint", value: 0 },
+        value: { a: null, tag: "num", value: BigInt(e.items.length) },
+      };
+      const assignsList: IR.Stmt<Annotation>[] = e.items.map((e, i) => {
+        const [init, stmt, vale] = flattenExprToVal(e, env);
+        inits = [...inits, ...init];
+        stmts = [...stmts, ...stmt];
+        return {
+          tag: "store",
+          start: { tag: "id", name: newListName },
+          offset: { tag: "wasmint", value: i + 1 },
+          value: vale,
+        };
+      });
+      return [
+        [
+          {
+            name: newListName,
+            type: e.a.type,
+            value: { a: e.a, tag: "none" },
+          },
+          ...inits,
+        ],
+        [
+          { tag: "assign", name: newListName, value: listAlloc },
+          ...stmts,
+          storeLength,
+          ...assignsList,
+        ],
+        {
+          a: e.a,
+          tag: "value",
+          value: {
+            a: e.a,
+            tag: "id",
+            name: newListName
+          },
+        },
+      ];
+    case "index":
+      var [objInit, objStmts, objExpr] = flattenExprToVal(e.obj, env);
+      var [idxInit, idxStmts, idxExpr] = flattenExprToVal(e.index, env);
+
+      var noneCheck: IR.Stmt<Annotation> = ERRORS.flattenAssertNotNone(e.a, objExpr);
+
+      // Bounds check code
+      var lenVar = generateName("listLen");
+      var lenStmt: IR.Stmt<Annotation> = {
+        a: {...idxExpr.a, type: {tag: "number"}},
+        tag: "assign",
+        name: lenVar,
+        value: {
+          a: {...idxExpr.a, type: {tag: "number"}},
+          tag: "load",
+          start: objExpr,
+          offset: {a: {...idxExpr.a, type: {tag: "number"}}, tag: "wasmint", value: 0}
+        }
+      };
+      var checkBoundStmt: IR.Stmt<Annotation> = ERRORS.flattenIndexOutOfBounds(e.a, idxExpr, {a: {...idxExpr.a, type: {tag: "number"}}, tag: "id", name: lenVar});
+      var indexAdd: AST.Expr<Annotation> = {
+        a: {...idxExpr.a, type: {tag: "number"}},
+        tag: "binop",
+        op: AST.BinOp.Plus,
+        left: e.index,
+        right: {a: {...idxExpr.a, type: {tag: "number"}}, tag: "literal", value: {tag: "num", value: 1}}
+      };
+      var [idxAddInit, idxAddStmts, idxAddExpr] = flattenExprToVal(indexAdd, env);
+
+      return [
+        [...objInit, ...idxInit, ...idxAddInit,
```

Comment:
I'm concerned about this, because it suggests that if we had something like `lst[f()]`, then we'd evaluate `f()` twice. Can you check on that?